source/common/common/inline_map_registry.h

```diff
@@ -0,0 +1,164 @@
+  template <class Scope> class InlineMapKeyRegister {
+  public:
+    InlineMapKeyRegister(absl::string_view);
+  };
+
+  using InlineMapDescriptor = InlineMapDescriptor<std::string>;
+  using Handle = InlineMapDescriptor::Handle;
+
+  struct ScopeInlineKeys {
+    std::string name;
+    absl::flat_hash_set<std::string> keys;
+  };
+
+  /**
+   * Scope inline keys vector type. This is used to finalize the registry and add inline keys to
+   * the descriptor based on the scope name.
+   */
+  using ScopeInlineKeysVector = std::vector<ScopeInlineKeys>;
+
+  /**
+   * Registry info type. This is used to get scope name and all managed descriptors' inline keys as
+   * string.
+   */
+  using RegistryInfo = std::vector<std::pair<std::string, std::string>>;
+
+private:
+  using Registry = absl::flat_hash_map<std::string, InlineMapDescriptor*>;
+
+  template <class Scope> friend class InlineMapDescriptorRegister;
+  template <class Scope> friend class InlineMapKeyRegister;
+
+  /**
+   * Call once flag to ensure the registry is finalized only once on multi-thread environment.
+   */
+  static std::once_flag& onceFlag() { MUTABLE_CONSTRUCT_ON_FIRST_USE(std::once_flag); }
+
+  /**
+   * Finalized flag to check whether the registry is finalized.
+   */
+  static bool& finalized() { MUTABLE_CONSTRUCT_ON_FIRST_USE(bool, false); }
+
+  /**
+   * A set of inline map registry that indexed by the scope name.
+   */
+  static Registry& registry() { MUTABLE_CONSTRUCT_ON_FIRST_USE(Registry); }
+
+  /**
+   * Create the inline map descriptor based on the scope name and insert it into the registry.
+   */
+  static InlineMapDescriptor* createDescriptor(absl::string_view scope_name);
+
+  /**
+   * Get or create the inline map registry singleton based on the scope type.
+   */
+  template <class Scope> static InlineMapDescriptor& getOrCreateDescriptor() {
+    // The function level static variable will be initialized only once and is thread safe.
+    // This is same with MUTABLE_CONSTRUCT_ON_FIRST_USE but has complex initialization.
+    static InlineMapDescriptor* initialize_once_descriptor = createDescriptor(Scope::name());
+    return *initialize_once_descriptor;
+  }
+
+  /**
+   * Add inline key to the descriptor based on the scope.
+   */
+  template <class Scope> static void addInlineKey(absl::string_view key) {
+    getOrCreateDescriptor<Scope>().addInlineKey(key);
+  }
+
+public:
+  /**
+   * Get the inline handle by the key name.
+   */
+  template <class Scope> static absl::optional<Handle> getHandleByKey(absl::string_view key) {
+    return getOrCreateDescriptor<Scope>().getHandleByKey(key);
+  }
+
+  /**
+   * Create inline map based on the scope and value type.
+   */
+  template <class Scope, class Value> static auto createInlineMap() {
+    return InlineMap<std::string, Value>::create(getOrCreateDescriptor<Scope>());
+  }
+
+  /**
+   * Get scope name and all managed descriptors' inline keys as string.
+   */
+  static RegistryInfo registryInfo();
```

Comment:
why copy the vector of string? Can you return it as a const ref?

Reply:
Basically this method should only be called few times at server instance(s) is/are initialized. It's performance doesn't matter.
And use copy semantics, we needn't considering the case where multiple server instances are initialized and the method is called from multiple threads at same time.